src/stepFunctions/commands/visualizeStateMachine.ts

```diff
@@ -29,257 +29,339 @@ export interface messageObject {
-            return
-        }
+    public deleteVisualization(visualizationToDelete: AslVisualization) {
+        this.managedVisualizations.delete(visualizationToDelete)
+    }
 
-        logger.debug('Sending update message to webview.')
+    private createNewVisualization(textDocument: vscode.TextDocument): vscode.WebviewPanel | void {
+        const newVisualization = new AslVisualization(textDocument)
+        this.managedVisualizations.add(newVisualization)
 
-        panel.webview.postMessage({
-            command: 'update',
-            stateMachineData: updatedTextDocument.getText(),
-            isValid,
+        newVisualization.getOnVisualizationDisposeEvent()(() => {
+            this.deleteVisualization(newVisualization)
         })
+
+        return newVisualization.getPanel()
     }
 
-    const debouncedUpdate = debounce(sendUpdateMessage, 500)
-
-    // Set the initial html for the webpage
-    panel.webview.html = getWebviewContent(
-        ext.visualizationResourcePaths.webviewBodyScript.with({ scheme: 'vscode-resource' }),
-        ext.visualizationResourcePaths.visualizationLibraryScript.with({ scheme: 'vscode-resource' }),
-        ext.visualizationResourcePaths.visualizationLibraryCSS.with({ scheme: 'vscode-resource' }),
-        ext.visualizationResourcePaths.stateMachineCustomThemeCSS.with({ scheme: 'vscode-resource' }),
-        {
-            inSync: localize(
-                'AWS.stepFunctions.graph.status.inSync',
-                'Previewing ASL document. <a href="" style="text-decoration:none;">View</a>'
-            ),
-            notInSync: localize('AWS.stepFunctions.graph.status.notInSync', 'Errors detected. Cannot preview.'),
-            syncing: localize('AWS.stepFunctions.graph.status.syncing', 'Rendering ASL graph...'),
+    private getExistingVisualization(documentToFind: vscode.TextDocument): AslVisualization | void {
+        for (const vis of this.managedVisualizations) {
+            if (vis.documentUri.path === documentToFind.uri.path) {
+                return vis
+            }
         }
-    )
 
-    // Add listener function to update the graph on document save
-    const updateOnSaveDisposable = vscode.workspace.onDidSaveTextDocument(async savedTextDocument => {
-        if (savedTextDocument && savedTextDocument.uri.path === documentUri.path) {
-            await sendUpdateMessage(savedTextDocument)
+        return
+    }
+}
+
+export class AslVisualization {
+    public readonly documentUri: vscode.Uri
+    public readonly webviewPanel: vscode.WebviewPanel
+    protected readonly disposables: vscode.Disposable[] = []
+    protected isPanelDisposed: boolean = false
+    private readonly onVisualizationDisposeEmitter = new vscode.EventEmitter<void>()
+
+    public constructor(textDocument: vscode.TextDocument) {
+        this.documentUri = textDocument.uri
+        this.webviewPanel = this.setupWebviewPanel(textDocument)
+    }
+
+    public getOnVisualizationDisposeEvent() {
```

Comment:
The event should be surfaced as a property, for example `public get onVisualizationDisposeEvent() {`
The usage is then `foo.onVisualizationDisposeEvent(() => myHandlerHere)`